lib/recursive-methods.rb

```diff
@@ -1,49 +1,87 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+  return 1 if n == 0 || n == 1
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse(s)
-    raise NotImplementedError, "Method not implemented"
-end
+# Time complexity: O(n)
+# Space complexity: O(1)
+def reverse(n)
+  i = 0
+  j = n.length - 1 
+  while i < j
+    temp = n[i]
+    n[i] = n[j]
+    n[j] = temp 
+    i += 1
+    j -= 1
+  end
+  return n 
+end 
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def reverse_inplace(s, i = 0, j = s.length - 1)
+  return s if i >= j
+  temp = s[i]
+  s[i] = s[j]
+  s[j] = temp 
+  return reverse_inplace(s, i+1, j-1) 
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if n == 0
+  return 2 if n == 1
+  return bunny(n - 1) + bunny(1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def nested(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def nested(s, i = 0, left = 0)
+  return left == 0 if i == s.length
+  if s[i] == '('
+    left += 1
+  else
+    if left > 0
+      left -= 1
+    else
+      return false
+    end
+  end
+  nested(s, i + 1, left)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(log2(n))
+# Space complexity: O(log2(n))
+def search(array, value, l = 0, r = array.length - 1)
```

Comment:
Because you are sorting with each recursive call, this is actually 

O(n log n * log n)